File: src/MeetingRoomII.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

/**
 Given an array of meeting time intervals consisting of start and end times [[s1,e1],[s2,e2],...] (si < ei), find the minimum number of conference rooms required.

 For example,
 Given [[0, 30],[5, 10],[15, 20]],
 return 2.
 */

public class MeetingRoomII {
    public static void main(String[] args) {

    }

    public int minMeetingRooms(Interval[] intervals) {
        if (intervals == null || intervals.length == 0) {
            return 0;
        }

        // create a comparator
        Interval[] temp = intervals.clone();

        Arrays.sort(temp, (a, b) -> (a.start - b.start));
        PriorityQueue<Interval> queue = new PriorityQueue<>((a, b) -> (a.end - b.end));

        int max = 0, n = intervals.length, i;

        for (i = 0; i < n; i++) {
            while (!queue.isEmpty() && queue.peek().end <= intervals[i].start) {
                queue.poll();
            }
            queue.offer(intervals[i]);
            max = Math.max(max, queue.size());
        }

        return max;
    }

    public class Interval {
        int start;
        int end;
        Interval() { start = 0; end = 0; }
        Interval(int s, int e) { start = s; end = e; }
    }
}

File: chronological-ordering/meeting-room-ii.ts
/**
Given an array of meeting time intervals intervals where intervals[i] = 
[starti, endi], return the minimum number of conference rooms required.

Leetcode: https://leetcode.com/problems/meeting-rooms-ii/

Example 1:

Input: intervals = [[0,30],[5,10],[15,20]]
Output: 2

Example 2:

Input: intervals = [[7,10],[2,4]]
Output: 1

Constraints:

1 <= intervals.length <= 104
0 <= starti < endi <= 106
*/
function minMeetingRooms(intervals: number[][]): number {
  const start = [];
  const end = [];

  for (let interval of intervals) {
    start.push(interval[0]);
    end.push(interval[1]);
  }

  start.sort((a, b) => a - b);
  end.sort((a, b) => a - b);
  let rooms = 0;
  let endIndex = 0;

  for (let i = 0; i < start.length; i++) {
    // whenever there is a start meeting, we need to add one room.
    //  But before adding rooms, we check to see if any previous meeting ends,
    //  which is why we check start with the first end. When the start is bigger
    //  than end, it means at this time one of the previous meeting ends,
    //  and it can take and reuse that room.
    rooms += 1;
    if (start[i] >= end[endIndex]) {
      rooms -= 1;
      endIndex += 1;
    }
  }
  return rooms;
}
